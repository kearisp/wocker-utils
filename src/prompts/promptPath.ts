import {
    createPrompt,
    useState,
    useMemo,
    useEffect,
    useKeypress,
    usePrefix,
    usePagination,
    isEnterKey,
    isTabKey,
    isUpKey,
    isDownKey,
    makeTheme,
    Status
} from "@inquirer/core";
import figures from "@inquirer/figures";
import {InquirerReadline} from "@inquirer/type";
import FS from "fs";
import Path from "path";
import {PromptConfig} from "../types/PromptConfig";
import {Theme} from "../types/Theme";
import {validatePrompt} from "../validation/validatePrompt";
import {prepareMessage} from "../tools/prepareMessage";
import {prepareValue} from "../tools/prepareValue";
import {prepareHelp} from "../tools/prepareHelp";


type Config = PromptConfig<string, {
    type?: "file" | "directory" | "any";
    basePath?: string;
    showHidden?: boolean;
}>;

export const promptPath = createPrompt<string, Config>((config: Config, done: (value: string) => void) => {
    const {
        message = "Enter path",
        type = "any",
        basePath = process.cwd(),
        showHidden = false,
        default: defaultValue = ""
    } = config;

    const theme = makeTheme<Theme>({
        style: {
            message: prepareMessage,
            value: prepareValue,
            help: prepareHelp
        }
    }, config.theme);
    const [inputValue, setInputValue] = useState(defaultValue),
          [status, setStatus] = useState<Status>("idle"),
          [error, setError] = useState(""),
          [selectedSuggestion, setSelectedSuggestion] = useState(0);
    const [base, currentDir] = useMemo((): [string, string] => {
        const fullPath = inputValue
            ? (Path.isAbsolute(inputValue) ? inputValue : Path.join(basePath, inputValue))
            : basePath;

        try {
            const stat = FS.statSync(fullPath);

            if(stat.isDirectory()) {
                return ["", fullPath];
            }
        }
        catch(err) {}

        return inputValue
            ? [Path.basename(fullPath), Path.dirname(fullPath)]
            : ["", fullPath];
    }, [inputValue]);
    const suggestions = useMemo((): string[] => {
        try {
            if(!FS.existsSync(currentDir)) {
                return [];
            }

            const files = FS.readdirSync(currentDir);

            return files
                .filter(file => {
                    if(!showHidden && file.startsWith("."))
                        return false;

                    if(!file.toLowerCase().startsWith(base.toLowerCase()))
                        return false;

                    const filePath = Path.join(currentDir, file);
                    const stats = FS.statSync(filePath);

                    if(type === "file")
                        return stats.isFile();

                    if(type === "directory")
                        return stats.isDirectory();

                    return true;
                })
                .sort((a, b) => {
                    const aPath = Path.join(currentDir, a);
                    const bPath = Path.join(currentDir, b);
                    const aIsDir = FS.statSync(aPath).isDirectory();
                    const bIsDir = FS.statSync(bPath).isDirectory();

                    if(aIsDir && !bIsDir)
                        return -1;

                    if(!aIsDir && bIsDir)
                        return 1;

                    return a.localeCompare(b);
                });
        }
        catch(err) {
            return [];
        }
    }, [inputValue]);

    const icon = usePrefix({ theme, status });

    const page = usePagination({
        items: suggestions,
        active: selectedSuggestion,
        pageSize: 6,
        renderItem: ({item, isActive}) => {
            const status = isActive
                ? "selected"
                : "idle";

            return [
                isActive
                    ? theme.style.value(figures.pointer, "active")
                    : " ",
                " ",
                theme.style.value(item, status)
            ].join("");
        }
    });

    const handleUp = (): void => {
        setSelectedSuggestion((suggestions.length + selectedSuggestion - 1) % suggestions.length);
    };

    const handleDown = (): void => {
        if(suggestions.length === 0) {
            return;
        }

        setSelectedSuggestion((selectedSuggestion + 1) % suggestions.length);
    };

    const handleChange = (readline: InquirerReadline): void => {
        setInputValue(readline.line);
        setSelectedSuggestion(0);
        setStatus("idle");
        setError("");
    };

    const handleTab = (readline: InquirerReadline): void => {
        if(suggestions.length === 0)
            return;

        const suggestion = suggestions[selectedSuggestion];
        const newPath = Path.join(currentDir, suggestion);
        const relativePath = Path.isAbsolute(inputValue)
            ? newPath
            : Path.relative(basePath, newPath);

        const stats = FS.statSync(newPath);
        const finalPath = stats.isDirectory()
            ? relativePath + Path.sep
            : relativePath;

        readline.clearLine(-1);
        readline.write(finalPath);
        setInputValue(finalPath);
        setSelectedSuggestion(0);
    };

    const handleSubmit = async (readline: InquirerReadline): Promise<void> => {
        setStatus("loading");

        const error = await validatePrompt(inputValue, config);

        if(error) {
            readline.write(inputValue);
            setStatus("error");
            setError(error.message);
            return;
        }

        setStatus("done");
        setError("");
        done(inputValue);
    };

    useEffect((): void => {
        setSelectedSuggestion(0);
    }, [inputValue]);

    useKeypress(async (key, readline): Promise<void> => {
        if(isEnterKey(key)) {
            return await handleSubmit(readline);
        }

        if(isTabKey(key)) {
            return handleTab(readline);
        }

        if(isUpKey(key)) {
            return handleUp();
        }

        if(isDownKey(key)) {
            return handleDown();
        }

        handleChange(readline);
    });

    return [
        [
            `${icon} `,
            theme.style.message(message, status),
            theme.style.value(inputValue, status)
        ].join(""),
        [
            error ? `${theme.style.error(error)}` : "",
            status !== "done" && page ? `\n${page}` : ""
        ].join("")
    ];
});
