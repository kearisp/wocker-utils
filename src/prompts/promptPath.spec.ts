import {describe, it, beforeEach, expect} from "@jest/globals";
import {render} from "@inquirer/testing";
import FS from "fs";
import {promptPath} from "./promptPath";


jest.mock("fs");

describe("promptPath", (): void => {
    const mockFS = FS as jest.Mocked<typeof FS>;

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("should return entered path on submit", async () => {
        mockFS.existsSync.mockReturnValue(true);
        mockFS.readdirSync.mockReturnValue([]);
        mockFS.statSync.mockReturnValue({
            isDirectory: () => false,
            isFile: () => true
        } as FS.Stats);

        const {answer, events, getScreen} = await render(promptPath, {
            message: "Enter file path"
        });

        expect(getScreen()).toMatchSnapshot();

        events.type("test.txt");
        events.keypress("enter");

        await expect(answer).resolves.toEqual("test.txt");
    });

    it("should show suggestions when typing", async () => {
        mockFS.existsSync.mockReturnValue(true);
        mockFS.readdirSync.mockReturnValue(["test1.txt", "test2.txt", "other.txt"] as any);
        mockFS.statSync.mockReturnValue({
            isDirectory: () => false,
            isFile: () => true
        } as FS.Stats);

        const {events, getScreen} = await render(promptPath, {
            message: "Enter file path"
        });

        events.type("test");

        const screen = getScreen();
        expect(screen).toContain("test1.txt");
        expect(screen).toContain("test2.txt");
        expect(screen).not.toContain("other.txt");
    });

    it("should autocomplete on tab", async () => {
        mockFS.existsSync.mockReturnValue(true);
        mockFS.readdirSync.mockReturnValue(["testfile.txt"] as any);
        mockFS.statSync.mockReturnValue({
            isDirectory: () => false,
            isFile: () => true
        } as FS.Stats);

        const {answer, events, getScreen} = await render(promptPath, {
            message: "Enter file path"
        });

        events.type("test");
        events.keypress("tab");
        events.keypress("enter");

        const screen = getScreen();
        expect(screen).toContain("testfile.txt");

        await expect(answer).resolves.toEqual("testfile.txt");
    });

    it("should navigate suggestions with arrows", async () => {
        mockFS.existsSync.mockReturnValue(true);
        mockFS.readdirSync.mockReturnValue(["file1.txt", "file2.txt", "file3.txt"] as any);
        mockFS.statSync.mockReturnValue({
            isDirectory: () => false,
            isFile: () => true
        } as FS.Stats);

        const {events, getScreen} = await render(promptPath, {
            message: "Enter file path"
        });

        events.type("file");
        events.keypress("down");
        events.keypress("down");

        const screen = getScreen();
        expect(screen).toMatchSnapshot();
    });

    it("should add separator for directories", async () => {
        mockFS.existsSync.mockReturnValue(true);
        mockFS.readdirSync.mockReturnValue(["testdir"] as any);
        mockFS.statSync.mockReturnValue({
            isDirectory: () => true,
            isFile: () => false
        } as FS.Stats);

        const {answer, events} = await render(promptPath, {
            message: "Enter directory path"
        });

        events.type("test");
        events.keypress("tab");
        events.keypress("enter");

        await expect(answer).resolves.toMatch(/testdir[\/\\]$/);
    });

    it("should filter by type (file only)", async () => {
        mockFS.existsSync.mockReturnValue(true);
        mockFS.readdirSync.mockReturnValue(["file.txt", "directory"] as any);
        mockFS.statSync.mockImplementation((path: any) => {
            const isDir = path.toString().includes("directory");
            return {
                isDirectory: () => isDir,
                isFile: () => !isDir
            } as FS.Stats;
        });

        const {events, getScreen} = await render(promptPath, {
            message: "Enter file path",
            filter: "directory"
        });

        events.type("");

        const screen = getScreen();

        expect(screen).toContain("directory");
        expect(screen).not.toContain("file.txt");
    });

    it("should hide hidden files by default", async () => {
        mockFS.existsSync.mockReturnValue(true);
        mockFS.readdirSync.mockReturnValue(["visible.txt", ".hidden"] as any);
        mockFS.statSync.mockReturnValue({
            isDirectory: () => false,
            isFile: () => true
        } as FS.Stats);

        const {events, getScreen} = await render(promptPath, {
            message: "Enter file path"
        });

        events.type("");

        const screen = getScreen();

        expect(screen).toContain("visible.txt");
        expect(screen).not.toContain(".hidden");
    });

    it("should show hidden files when enabled", async () => {
        mockFS.existsSync.mockReturnValue(true);
        mockFS.readdirSync.mockReturnValue([".hidden", "visible.txt"] as any);
        mockFS.statSync.mockReturnValue({
            isDirectory: () => false,
            isFile: () => true
        } as FS.Stats);

        const {events, getScreen} = await render(promptPath, {
            message: "Enter file path",
            showHidden: true
        });

        events.type("");

        const screen = getScreen();

        expect(screen).toContain(".hidden");
        expect(screen).toContain("visible.txt");
    });
});
