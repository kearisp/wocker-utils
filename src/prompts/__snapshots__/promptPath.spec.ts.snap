// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`promptPath should navigate suggestions with arrows 1`] = `
"? Enter file path: file

  file1.txt
  file2.txt
‚ùØ file3.txt"
`;

exports[`promptPath should return entered path on submit 1`] = `"? Enter file path:"`;
